x/concentrated-liquidity/fees.go

```diff
@@ -189,46 +190,62 @@ func (k Keeper) collectFees(ctx sdk.Context, poolId uint64, owner sdk.AccAddress
 		return sdk.Coins{}, err
 	}
 
-	positionKey := formatPositionAccumulatorKey(poolId, owner, lowerTick, upperTick)
-
-	hasPosition, err := feeAccumulator.HasPosition(positionKey)
+	// There may be multiple positions with the same lower and upper tick, but different frozenUntil times.
+	// We retrieve all of them and claim fees for each one.
+	positions, err := k.getAllPositionsWithVaryingFreezeTimes(ctx, poolId, owner, lowerTick, upperTick)
 	if err != nil {
 		return sdk.Coins{}, err
 	}
 
-	if !hasPosition {
+	if len(positions) == 0 {
 		return sdk.Coins{}, cltypes.PositionNotFoundError{PoolId: poolId, LowerTick: lowerTick, UpperTick: upperTick}
 	}
 
-	// compute fee growth outside of the range between lower tick and upper tick.
-	feeGrowthOutside, err := k.getFeeGrowthOutside(ctx, poolId, lowerTick, upperTick)
-	if err != nil {
-		return sdk.Coins{}, err
-	}
+	var totalFeesClaimed sdk.Coins
 
-	// We need to update the position's accumulator to the current fee growth outside
-	// before we claim rewards.
-	if err := feeAccumulator.SetPositionCustomAcc(positionKey, feeGrowthOutside); err != nil {
-		return sdk.Coins{}, err
-	}
+	for _, position := range positions {
+		positionKey := formatPositionAccumulatorKey(poolId, owner, lowerTick, upperTick, position.FrozenUntil)
 
-	// claim fees.
-	feesClaimed, err := feeAccumulator.ClaimRewards(positionKey)
-	if err != nil {
-		return sdk.Coins{}, err
+		hasPosition, err := feeAccumulator.HasPosition(positionKey)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+
+		if !hasPosition {
+			return sdk.Coins{}, cltypes.PositionNotFoundError{PoolId: poolId, LowerTick: lowerTick, UpperTick: upperTick}
+		}
+
+		// compute fee growth outside of the range between lower tick and upper tick.
+		feeGrowthOutside, err := k.getFeeGrowthOutside(ctx, poolId, lowerTick, upperTick)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+
+		// We need to update the position's accumulator to the current fee growth outside
+		// before we claim rewards.
+		if err := feeAccumulator.SetPositionCustomAcc(positionKey, feeGrowthOutside); err != nil {
+			return sdk.Coins{}, err
+		}
+
+		// claim fees.
+		feesClaimed, err := feeAccumulator.ClaimRewards(positionKey)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+
+		// Update total fees claim counter
+		totalFeesClaimed = totalFeesClaimed.Add(feesClaimed...)
 	}
 
+	// Once we have iterated through all the positions, we do a single bank send from the pool to the owner.
 	pool, err := k.getPoolById(ctx, poolId)
 	if err != nil {
 		return sdk.Coins{}, err
 	}
-
-	// distribute the fees from pool to the position owner.
-	if err := k.bankKeeper.SendCoins(ctx, pool.GetAddress(), owner, feesClaimed); err != nil {
+	if err := k.bankKeeper.SendCoins(ctx, pool.GetAddress(), owner, totalFeesClaimed); err != nil {
```

Comment:
Note to reviewer: I decided to do a single bank send after iterating over all positions and claiming the rewards in order to reduce the overhead. Please lmk if you agree / can think of a way to do this with even less overhead.